memory/jemalloc_nodump_allocator.cc

```diff
@@ -106,45 +135,66 @@ void JemallocNodumpAllocator::Deallocate(void* p) {
+    arena_indexes_.emplace_back();
+    size_t arena_index_size = sizeof(arena_indexes_.back());
+    int ret = mallctl("arenas.create", &arena_indexes_.back(),
+                      &arena_index_size, nullptr, 0);
+    if (ret != 0) {
+      // API doc says no arena is created in failure case, so we don't need to
+      // remember it for future cleanup.
+      arena_indexes_.pop_back();
+      return Status::Incomplete(
+          "Failed to create jemalloc arena, error code: " +
+          std::to_string(ret));
+    }
+
+    // Read existing hooks.
+    std::string key =
+        "arena." + std::to_string(arena_indexes_[i]) + ".extent_hooks";
+    extent_hooks_t* hooks;
+    size_t hooks_size = sizeof(hooks);
+    ret = mallctl(key.c_str(), &hooks, &hooks_size, nullptr, 0);
+    if (ret != 0) {
+      return Status::Incomplete("Failed to read existing hooks, error code: " +
+                                std::to_string(ret));
+    }
+
+    // Store existing alloc.
+    extent_alloc_t* original_alloc = hooks->alloc;
+    extent_alloc_t* expected = nullptr;
+    bool success =
+        JemallocNodumpAllocator::original_alloc_.compare_exchange_strong(
+            expected, original_alloc);
+    if (!success && original_alloc != expected) {
+      // This could happen if jemalloc creates new arenas with different initial
+      // values in their `alloc` function pointers. See `original_alloc_` API
+      // doc for more details.
+      return Status::Incomplete("Original alloc conflict.");
+    }
 
-  // Store existing alloc.
-  extent_alloc_t* original_alloc = hooks->alloc;
-  extent_alloc_t* expected = nullptr;
-  bool success =
-      JemallocNodumpAllocator::original_alloc_.compare_exchange_strong(
-          expected, original_alloc);
-  if (!success && original_alloc != expected) {
-    return Status::Incomplete("Original alloc conflict.");
+    // Set the custom hook.
+    per_arena_hooks_.emplace_back();
+    per_arena_hooks_.back().reset(new extent_hooks_t(*hooks));
+    per_arena_hooks_.back()->alloc = &JemallocNodumpAllocator::Alloc;
+    extent_hooks_t* hooks_ptr = per_arena_hooks_.back().get();
+    ret = mallctl(key.c_str(), nullptr, nullptr, &hooks_ptr, sizeof(hooks_ptr));
+    if (ret != 0) {
+      return Status::Incomplete("Failed to set custom hook, error code: " +
+                                std::to_string(ret));
+    }
   }
 
-  // Set the custom hook.
-  arena_hooks_.reset(new extent_hooks_t(*hooks));
-  arena_hooks_->alloc = &JemallocNodumpAllocator::Alloc;
-  extent_hooks_t* hooks_ptr = arena_hooks_.get();
-  ret = mallctl(key.c_str(), nullptr, nullptr, &hooks_ptr, sizeof(hooks_ptr));
-  if (ret != 0) {
-    return Status::Incomplete("Failed to set custom hook, error code: " +
-                              std::to_string(ret));
+  log2_num_arenas_ = 0;
+  while ((1ul << log2_num_arenas_) < arena_indexes_.size()) {
+    log2_num_arenas_++;
   }
+  assert((1ul << log2_num_arenas_) == arena_indexes_.size());
```

Comment:
This should not be an assertion, it is not an internal logic error, it is an invalid input provided by the user and so an error should be returned.

But I would just relax the requirement: if the RNG has good entropy on the high bits, we don't really need a power of 2 here, we can use a different range reduction algorithm:

```
uint32_t arena_index = (static_cast<uint64_t>(num_arenas) * static_cast<uint32_t>(tl_random.Next()) >> 32;
```
which is almost as fast as a binary mask.

Reply:
The error is returned by `PrepareOptions()`; this assertion would only fail if that validation failed to prevent invalid arguments from reaching here, which seems like an internal logic error.

Reply:
Ah! I missed the validation code.

Reply:
I would still suggest to allow arbitrary `num_arenas`, it costs almost nothing and we don't need to take care of invalid configuration.

Reply:
Sure, done.